### ISSUE TITLE
docker.service tops the systemd-analyze critical path

### ISSUE THREAD

**Issue by mvdan** [moby/moby#38373]
This is on systemd 239.303, and the Docker version below:

```
Server:
 Engine:
  Version:          18.09.0-ce
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.11.2
  Git commit:       4d60db472b
  Built:            Fri Nov  9 00:05:11 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

I'd like to have Docker start up during boot, as otherwise I need to remember to `systemctl start docker` anytime I need it - and after all, I end up using it most days. However, getting it started takes about two seconds, which are spent blocking systemd from showing me any UI at the end of the boot. Below are some `systemd-analyze` numbers from the most recent boot.

```
$ systemd-analyze blame
          1.974s docker.service
           475ms systemd-modules-load.service
           299ms netctl-auto@wlp2s0.service
           259ms tlp.service
           233ms dev-nvme0n1p2.device
           226ms ldconfig.service
            67ms systemd-journal-flush.service
            56ms systemd-udev-trigger.service
            50ms boot.mount
            49ms systemd-udevd.service
            43ms systemd-journald.service
            39ms systemd-sysctl.service
            38ms polkit.service
            34ms systemd-logind.service
            32ms user@1000.service
            21ms systemd-sysusers.service
            15ms systemd-remount-fs.service
            15ms systemd-tmpfiles-setup.service
            13ms systemd-journal-catalog-update.service
            10ms systemd-update-done.service
             9ms systemd-backlight@leds:tpacpi::kbd_backlight.service
             8ms systemd-update-utmp.service
             8ms systemd-user-sessions.service
             7ms dev-mqueue.mount
             7ms kmod-static-nodes.service
             7ms dev-hugepages.mount
             6ms systemd-tmpfiles-setup-dev.service
             6ms sys-kernel-debug.mount
             5ms user-runtime-dir@1000.service
             5ms systemd-random-seed.service
             3ms rtkit-daemon.service
             3ms sys-kernel-config.mount
             2ms tmp.mount
             1ms docker.socket
$ systemd-analyze critical-chain
The time after the unit is active or started is printed after the "@" character.
The time the unit takes to start is printed after the "+" character.

graphical.target @2.709s
└─multi-user.target @2.706s
  └─docker.service @723ms +1.974s
    └─basic.target @696ms
      └─sockets.target @695ms
        └─docker.socket @692ms +1ms
          └─sysinit.target @691ms
            └─systemd-update-done.service @680ms +10ms
              └─ldconfig.service @452ms +226ms
                └─local-fs.target @449ms
                  └─var-lib-docker-btrfs.mount @2.696s
                    └─dev-nvme0n1p2.device @163ms +233ms
```

In the first command, one can see that `docker.service` is by far the service that takes the most time to start. But that's not necessarily a problem; for example, `tlp.service` takes 250ms, but it doesn't show up in the critical-chain command, hence it's not slowing down how long it takes from me pushing the power button to seeing my desktop.

It would be great if `docker.service` was asynchronous in the same way. Forgive me if the terminology is inexact; I'm not an expert on Systemd nor in writing service files. But it sounds to me like `graphical.target` should not be blocked by `docker.service`. The cause seems to be that `graphical.target` depends on `sockets.target`, which waits for all sockets to be available, and then `docker.socket` waits for `docker.service` to be ready.

I realise that https://github.com/moby/moby/issues/34500 was raised before, but that was about the service taking an unusual amount of time to start. I think two seconds is reasonable, but only if it's not blocking my graphical user interface from showing up.

**Comment by mvdan** [comment]
> The cause seems to be

Ignore that sentence; I think I misread the critical path graph.

**Comment by thaJeztah** [comment]
Ah, yes, the `docker.socket` service provides socket activation, which will start the docker daemon when the socket is accessed.

Socket activation is useful for desktop situations, but for server setups, you'll likely want the service to be up on reboot.

Your use case may be something that we should document better though.

Not sure if there's a different workaround that provides the service to be up, without blocking the UI due to the socket not being enabled (other than a drop-in file to disable socket activation) 🤔

**Comment by mireczatko** [comment]
same here except that it takes extremely longer time 
└─docker.service @24.590s +11.538s

**Comment by cpuguy83** [comment]
Do you have some logs?
journalctl -u docker

**Comment by MPeti1** [comment]
It seems it's welcome to bring similar issues here, so I wouldn't open a new one for now, but feel free to tell me if I should.

On my system `docker.service` adds 34 seconds to the boot time, based on `systemd-analyze critical-chain`'s [output](https://gist.github.com/MPeti1/8907d52abf2e545447736a07bff5879e#file-systemd-analyze-critical-chain).
[Here](https://gist.github.com/MPeti1/8907d52abf2e545447736a07bff5879e#file-journalctl-u-docker-with-docker-debug-mode-on) you can find docker logs from right after a system reboot.

**Comment by mvdan** [comment]
I solved my problem via https://github.com/moby/moby/issues/38373#issuecomment-447393517, so I'm closing this for now. If others need to use docker.service and it's slow, I would suggest to file a new issue with details.

**Comment by tech3br** [comment]
> Ah, yes, the `docker.socket` service provides socket activation, which will start the docker daemon when the socket is accessed.
> 
> Socket activation is useful for desktop situations, but for server setups, you'll likely want the service to be up on reboot.
> 
> Your use case may be something that we should document better though.
> 
> Not sure if there's a different workaround that provides the service to be up, without blocking the UI due to the socket not being enabled (other than a drop-in file to disable socket activation) 🤔

useful information. Thank you!
